site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -0,0 +1,95 @@
+subtext: Good thing that shield flashing bug still isn't around
+
+How is there such a major difference in speed from the first and the second video? The speed difference boils down to how the 3DS GPU, the PICA200, is being emulated. The first video is showing off the software renderer, which emulates the PICA200 by using your host computer's CPU; on the other hand, the second video is using the OpenGL hardware renderer, which emulates the PICA200 by using your host computer's GPU. From those videos, using the GPU to emulate the 3DS GPU is the clear winner when it comes to speed! However, its not all sunshine and daisies; there's always tradeoffs in software development.
+
+## Challenges in the Hardware Renderer
+
+Earlier it was stated that the OpenGL hardware renderer was emulating the PICA200 by using the GPU instead of the CPU, and ... that's only partially true. As it stands, only a portion of the PICA200 emulation is running on the GPU; most of it is running on the CPU! To understand why, we need to understand a little about the difference between CPUs and modern GPUs.
+
+image: what parts of the gpu citra emulated on the gpu
+subtext: For a GPU hardware renderer, Citra isn't using the GPU much!
+
+As a general rule of thumb, CPUs are fast at computing general tasks, while GPUs are *blazing fast* at computing very specific tasks. Whenever the tasks that the PICA200 performs matches well with tasks that you can do on a GPU using OpenGL, everything is fast and everyone is happy! But far too often, we ran into edge cases that the PICA200 would run but, frankly, OpenGL is not well suited to support.
+
+picture: pokemon outline bug
+subtext: The infamous pokemon outlines bug was one such example of something that OpenGL just doesn't support
+
+When the hardware renderer was originally written, there was no other choice to use besides OpenGL. On the surface, OpenGL is a good candidate for a hardware renderer, as it runs on all 3 major desktop platforms. As long as we stay with the [8 year old OpenGL 3.3 standard](https://www.khronos.org/opengl/wiki/History_of_OpenGL#OpenGL_3.3_.282010.29), we can support a very wide range of hardware as well. But OpenGL can be very finicky; since its just a standard, there are several different OpenGL implementations in the real world. Its performance and features can vary widely between operating systems and graphics card manufacturer, and as you might guess, this leads to some [OS specific bugs that are very hard to track down](https://github.com/citra-emu/citra/issues/2416). In the linked issue, only on Mac OSX, Citra would leak memory from the hardware renderer. We traced it back to how textures were juggled between the 3DS memory and the host GPU, but we don't have many developers that use Mac, so we never did find what the root issue was. For a little bit of good news, this is fixed in the latest nightly, but only because the entire texture handling code was rewritten!
```

Comment:
This whole paragraph makes it sound like we're getting rid of OpenGL, might make people confused and think we're switching to Vulkan. I also don't really see how the MacOS memory leak is related

Reply:
The ending sentence is why "For a little bit of good news, this is fixed in the latest nightly, but only because the entire texture handling code was rewritten!"

Which leads into the next paragraph about the texture cache rewrite